numba/array.py

```diff
@@ -0,0 +1,177 @@
+from numba.tracing import trace
+import numpy
+import sys
+
+def _o2s(dtype, shape, order):
+    # convert order parameter to strides
+
+    if dtype is None or shape is None or order is None:
+        return None
+
+    if order == 'F':
+        shape = list(shape)
+        shape.reverse()
+    strides = []
+    itemsize = dtype.itemsize
+    for i in range(len(shape), 0, -1):
+        strides.append(itemsize)
+        itemsize *= shape[i - 1]
+    if order in ('C', None):
+        strides.reverse()
+    return tuple(strides)
+
+def _s2o(dtype, shape, strides):
+    # convert strides parameter to order
+    # Note: strides must correspond to contiguous data layout
+
+    if strides is None or strides[-1] == dtype.itemsize:
+        order = 'C'
+    elif strides[0] == dtype.itemsize:
+        order = 'F'
+    else:
+        raise Error('strides do not correspond to contiguous data layout')
+    s2 = _o2s(dtype, shape, order)
+    if strides != s2:
+        raise Error('strides do not correspond to contiguous data layout')
+    return order
+
+class Array:
+    """An array type that supports host and GPU storage."""
+
+    def __init__(self, obj=None, copy=True,
+                 shape=None, dtype=None, order=None, where='host'):
+        """Construct an Array in the memory space defined by 'where'."""
+
+        assert where in ('host', 'gpu')
+        # we need either a prototype or proper type info
+        assert obj is not None or (shape and dtype)
+        self._host = self._gpu = None
+        self._host_valid = self._gpu_valid = False
+        self._allocate(where, obj, dtype, shape, _o2s(dtype, shape, order), copy)
+        if where == 'host':
+            self._host_valid = True
+            t = self._host
+        else:
+            self._gpu_valid = True
+            t = self._gpu
+        self.shape = t.shape
+        self.strides = t.strides
+        self.dtype = t.dtype
+        self.ndim = t.ndim
+        self.size = t.size
+
+    def host(self):
+        """Get a host array."""
+
+        self._sync('host')
+        #self._invalidate('gpu') # only if write-access is requested !
+        return self._host
+
+    def host_changed(self):
+        """Mark the host array as changed, broadcast updates if needed."""
+
+        self._invalidate('gpu')
+        # only sync if there are active views
+        if self._gpu is not None and sys.getrefcount(self._gpu) > 2:
+            self._sync('gpu')
+
+    def gpu(self):
+        """Get a GPU array."""
```

Comment:
`Get a` kind of implies that there's some creation mechanism present, is it not the case that this is obtaining the array but in GPU memory (same applies to `.host()`? These are more like accessors ?